File: Java/DesignPatterns/src/main/java/designPatterns/behavioral/observer/Topic.java
package designPatterns.behavioral.observer;

import java.util.ArrayList;
import java.util.List;

public class Topic implements Subject {
  private List<Observer> observerList;
  private String message;

  public Topic() {
    observerList = new ArrayList<>();
  }

  @Override
  public void register(Observer o) {
    observerList.add(o);
  }

  @Override
  public void unregister(Observer o) {
    observerList.remove(o);
  }

  @Override
  public void notifyObservers() {
    for (Observer o: observerList) {
      o.update(message);
    }
  }

  public String postMessage(String msg) {
    message = msg;
    notifyObservers();
    return "Message Posted to Topic: " + msg;
  }
}

File: Kotlin/DesignPatterns/src/main/kotlin/behavioral/observer/Topic.kt
package behavioral.observer

class Topic : Subject {
  private lateinit var observerList: MutableList<Observer>
  private var message: String? = null

  init {
    observerList = ArrayList()
  }

  override fun register(o: Observer) {
    observerList.add(o)
  }

  override fun unregister(o: Observer) {
    observerList.remove(o)
  }

  override fun notifyObserver() {
    for (o in observerList) {
      o.update(message!!)
    }
  }

  fun postMessage(msg: String): String {
    message = msg
    notifyObserver()
    return "Message Posted to Topic: $msg"
  }
}